ocaml/xapi/xapi_pool.ml

```diff
@@ -781,15 +781,33 @@ let eject ~__context ~host =
 
 	if Pool_role.is_master () then raise Cannot_eject_master
 	else begin
-		(* Fail the operation if any VMs are running here (except control domains) *)
+		(* Fail the operation if any VMs are running here (except dom0) *)
 		let my_vms_with_records = Db.VM.get_records_where ~__context ~expr:(Eq(Field "resident_on", Literal (Ref.string_of host))) in
 		List.iter (fun (_, x) -> 
-			if (not x.API.vM_is_control_domain) && x.API.vM_power_state<>`Halted
+			if (not (Helpers.is_domain_zero ~__context (Db.VM.get_by_uuid ~__context ~uuid:x.API.vM_uuid))) && x.API.vM_power_state <>`Halted
 			then begin
 				error "VM uuid %s not in Halted state and resident_on this host" (x.API.vM_uuid);
 				raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["VM resident on host"]))
 			end) my_vms_with_records;
 
+		(* all control domains resident on me should be destroyed once I leave the
```

Comment:
This code should make the predicate more explicit when a control domain can be destroyed along the following lines. A List.filter takes a predicate as argument and it should be expressed as logical expressions and not using if-then-else.

    let vm_has_some_complex_property vm =
      ...

    let is_obsolete (vm_ref, vm_ref) =
      vm_rec.API.vM_is_control_domain
      || (Db.is_valid_ref __context vm_rec.API.vM_resident_on
        && vm_rec.API.vM_resident_on = host)
      || vm_has_some_complex_property vm

    let control_domains_to_destroy =
      List.filter is_obsolete (Db.VM.get_all_records ~__context)